src/Evolution/Systems/ScalarWave/Constraints.cpp

```diff
@@ -0,0 +1,97 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/ScalarWave/Constraints.hpp"
+
+#include <cstddef>
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+namespace ScalarWave {
+template <size_t SpatialDim>
+tnsr::i<DataVector, SpatialDim, Frame::Inertial> one_index_constraint(
+    const tnsr::i<DataVector, SpatialDim, Frame::Inertial>& d_psi,
+    const tnsr::i<DataVector, SpatialDim, Frame::Inertial>& phi) noexcept {
+  tnsr::i<DataVector, SpatialDim, Frame::Inertial> constraint(
+      get_size(get<0>(phi)));
+  one_index_constraint(make_not_null(&constraint), d_psi, phi);
+  return constraint;
+}
+
+template <size_t SpatialDim>
+void one_index_constraint(
+    const gsl::not_null<tnsr::i<DataVector, SpatialDim, Frame::Inertial>*>
+        constraint,
+    const tnsr::i<DataVector, SpatialDim, Frame::Inertial>& d_psi,
+    const tnsr::i<DataVector, SpatialDim, Frame::Inertial>& phi) noexcept {
+  if (get_size(get<0>(*constraint)) != get_size(get<0>(phi))) {
+    *constraint =
+        tnsr::i<DataVector, SpatialDim, Frame::Inertial>(get_size(get<0>(phi)));
```

Comment:
I think @moxcodes added a destructive-resize function a while back. Can you use it here? I find `destructive_resize_components` in `Utilities/ContainerHelpers.hpp`, that's probably the right one to use.

Same in the two-index constraint

Reply:
this is useful, thanks for the suggestion :+1: